### ISSUE TITLE
Changing the License

### ISSUE THREAD

**Issue by eudoxia0** [austral/austral#421]
tl;dr I'd like to change the license from GPLv3 to Apache 2.0 with the LLVM exception. I think this is good. This involves asking all the contributors for consent. See below for the rationale.

(Note that everything here is my layman's best-effort understanding.)

## Rationale

The compiler is GPLv3. I intended this to apply just to the compiler, and not the runtime, but since this was a personal project, I never got around to writing something like [GCC's Runtime Linking Exception][rle], or specific  licenses for the runtime, to ensure compiled output doesn't have to be GPL. I opened a few issues (#390, #409, #410) to relicense the runtime, builtins, and standard library, but I'm not sure if this is enough to comply with the wording of the GPL around linking.

[rle]: https://www.gnu.org/licenses/gcc-exception-3.1.html

My concern with this is that users will be put off by the inherent legal ambiguity of whether every program they write has to be GPL, since it gets compiled along with the runtime.

Needless to say, this has been a _huge_ headache. And, rather than untangle all the legal language, I've chosen to simply cut the knot: relicense the whole thing under a permissive license, so nobody has to be concerned about anything.

Additionally, I am not a lawyer, so I'd rather stick to the well-trod path.

## Considerations

When considering a license, I looked for the following:

1. The license should be **permissive**, to encourage contributions.

2. It should be **non-viral**, so there's no chance compiled programs may be subject to license restrictions.

3. It should be **well-understood**, not obscure or written by me.

4. It should have a **patent clause** to protect users from patent litigation.

5. It should be **uniform** between the compiler and the runtime.
    1. The reason for this is that, as the [LLVM project discovered][llvm-relicense], when your compiler and runtime have different licenses, you can't move freely from one to the other.

[llvm-relicense]: https://foundation.llvm.org/docs/relicensing/

## Choice of License

The choice came down to [Apache 2.0][ap] vs. [MIT][mit].

[ap]: https://en.wikipedia.org/wiki/Apache_License
[mit]: https://en.wikipedia.org/wiki/MIT_License

1. **Apache 2.0** is a permissive, widely-used, well-understood license. It has a patent clause, but it is incompatible with GPLv2. That is, code using Apache 2.0 can't be imported into GPLv2 programs. The largest GPLv2 ecosystem is the Linux kernel, so this is an important consideration. 
1. The **MIT License** is very widely used, and very permissive. It has no patent clause, but it is compatible with GPLv2.

[Rust is dual-licensed][rust] as MIT and Apache 2.0, and [Racket][racket] followed suit. My objection to dual licenses is it's necessarily more verbose, and it's not always obvious to users what dual licensing means.

[rust]: https://github.com/rust-lang/rust/blob/dffea43fc1102bdfe16d88ed412c23d4f0f08d9d/COPYRIGHT
[racket]: https://github.com/racket/racket/blob/master/LICENSE.txt

Fortunately, there already exists a license that satisfies all these requirements!

This is the [Apache 2.0 with LLVM Exceptions][apllvm] license. Which I admit is a bit of a mouthful. Here's a [paper][paper] about it. This variant of Apache 2.0 is both compatible with GPLv2 and suitable as a license for compilers. Additionally, it has apparently been looked at extensively by lawyers from both LLVM and the FSF. It's Apache 2.0 with the following two paragraphs:

[apllvm]: https://llvm.org/LICENSE.txt

[paper]: https://www.osti.gov/biblio/1782513

>As an exception, if, as a result of your compiling your source code, portions of this Software are embedded into an Object form of such source code, you may redistribute such embedded portions in such Object form without complying with the conditions of Sections 4(a), 4(b) and 4(d) of the License.
>
>In addition, if you combine or link compiled forms of this Software with software that is licensed under the GPLv2 ("Combined Software") and if a court of competent jurisdiction determines that the patent provision (Section 3), the indemnity provision (Section 9) or other Section of the License conflicts with the conditions of the GPLv2, you may retroactively and prospectively choose to deem waived or otherwise exclude such Section(s) of the License, but only in their entirety and only with respect to the Combined Software.

Basically compiler output has no attribution requirements, and there's a specific carve-out for GPLv2 compatibility. Win-win!

To be clear, the [SPDX][spdx] identifier for this license is: `Apache-2.0 WITH LLVM-exception`. The text of the Apache 2.0 license is [here](https://spdx.org/licenses/Apache-2.0.html), and the text of the LLVM exception is [here](https://spdx.org/licenses/LLVM-exception.html).

[spdx]: https://spdx.dev/

## Why not GCC's Runtime Library Exception?

GCC is GPLv3, but it has a Runtime Library Exception so that compiler output is not also GPL, that is, you can compile proprietary or otherwise-licensed programs with it. I'm not sure that I can copy-paste the linking exception into Austral without altering the wording. Some of the wording, which is confusing to me, seems like it might not apply to a compiler that emits high level language code (section 0 paragraph 4 particularly).

The point is that I'm not a lawyer. I'm not qualified to write an exception to a license, and getting it wrong has serious consequences.

## Why not two licenses?

The compiler could be GPL with a linking exception, and the runtime could be permissively licensed. There's two problems with this:

1. **Mobility:** as a compiler evolves, code naturally moves from the compiler to the runtime and vice-versa. For example, the compiler might emit a utility function which is then moved as a static function in the runtime. With two licenses, this is not possible (code can move from the permissively-licensed runtime to the GPL compiler, but not vice-versa). This isn't theoretical: [LLVM had this problem][llvm-relicense] because core LLVM and the runtime had different licenses. In their words, one of the goals of relicensing was to "Eliminate the schism between runtime libraries and the rest of the compiler that makes it difficult to move code between them".

2. **Fragility:** having multiple licenses involves drawing an invisible line that meanders across your codebase. If some code accidentally crosses the line, it might unwittingly make the whole thing GPL. This isn't great for the users, who expected one thing and got another.

## Why now?

Because [Racket's relicensing effort][racket-effort] took three years, since it involved reaching out to hundreds of people. [LLVM's relicensing effort is ongoing][llvm-effort] and has been since approximately 2017. By contrast, this repository only has eight contributors. It is tractable to reach out to everyone.

The best time to fix this was yesterday, the second best time is now.

[racket-effort]: https://github.com/racket/racket/issues/1570
[llvm-effort]: https://blog.llvm.org/posts/2021-11-18-relicensing-update/

## Prior Art

LLVM is moving to Apache 2.0 with the LLVM exception: https://llvm.org/LICENSE.txt

Microsoft's implementation of the C++ STL also uses this: https://github.com/microsoft/STL

Swift seems to use Apache 2.0 with just the first paragraph of the LLVM exception: https://github.com/apple/swift/blob/f1051f1520b9e51484e86a986985888b3d96385b/LICENSE.txt

## Summary

So, I'd like to ask anyone who's contributed: if you agree with relicensing under the Apache 2.0 license with the LLVM exception, would you kindly reply with

>I, [full name], hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

In the above, "the Austral project" means the repository at https://github.com/austral/austral.

If you don't agree, feel free to express your concerns [privately to me](mailto:fernando@borretti.me) or in this thread. Contributors hold the copyright to their contributions: you have a right to say no.

Additionally, if you have privacy concerns about giving your full name, please ask me and I will figure it out. But looking at the git logs it seems everyone's using their names.

## Contributors

Here's the list of all contributors (other than me) at the time of writing:

- @ishehadeh
- @qbradley
- @jacereda
- @AlskaPark
- @samueldurantes
- @AgostonSzepessy
- @scolsen
- @tekknolagi

(With huge apologies to @jacereda for pinging you on this topic for like, the fourth time. Hopefully this time I got it right.)

**Comment by tekknolagi** [comment]
I, Max Bernstein, hereby release all of my contributions to the Austral
project under the Apache Software License, version 2.0, with the LLVM
exception.

**Comment by qbradley** [comment]
I, Quetzal Bradley, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by jacereda** [comment]
I, Jorge Acereda, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by ishehadeh** [comment]
I, Ian Shehadeh, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by scolsen** [comment]
I, Scott Olsen, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by AgostonSzepessy** [comment]
I, Agoston Szepessy, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by AlskaPark** [comment]
I, Jaeyoung Park, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by eudoxia0** [comment]
Thanks everyone. cc @samueldurantes ICYMI.

**Comment by zetah11** [comment]
I, zetah11, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by samueldurantes** [comment]
I, Samuel Durante, hereby release all of my contributions to the Austral project under the Apache Software License, version 2.0, with the LLVM exception.

**Comment by eudoxia0** [comment]
Thanks everyone! This is fixed by #435
